File: julia/2022/day_07.jl
const total_disk_space = 70_000_000
const update_size = 30_000_000

struct File
    size::UInt
end

struct Dir
    parent::Union{Dir, Nothing}
    contents::Dict{String, Union{Dir, File}}
end

size(file::File)::UInt = file.size
size(dir::Dir)::UInt = values(dir.contents) .|> size |> sum

function main()
    text = read("input/2022/input_07.txt", String)
    cwd = root = Dir(nothing, Dict())
    all_dirs = [root]
    for line in eachsplit(text, '\n', keepempty=false)
        cd, size, name = match(r"ls|cd (/|\.\.|\w+)|(dir|\d+) ([\w.]+)", line)
        if !isnothing(cd)
            if cd == "/"
                cwd = root
            elseif cd == ".."
                cwd = cwd.parent
            else
                cwd = cwd.contents[cd]
            end
        elseif !isnothing(size)
            if size == "dir"
                dir = Dir(cwd, Dict())
                cwd.contents[name] = dir
                push!(all_dirs, dir)
            else
                cwd.contents[name] = File(parse(UInt, size))
            end
        end
    end
    small_sum = sum(filter(x -> x <= 100_000, size.(all_dirs)))
    println("Part 1: $small_sum")
    free_space = total_disk_space - size(root)
    space_needed = update_size - free_space
    space_to_free = minimum(filter(x -> x >= space_needed, size.(all_dirs)))
    println("Part 2: $space_to_free")
end

main()

File: rust/2022/src/bin/day_07.rs
use std::collections::HashMap;

const DIR_MAX_SIZE: u64 = 100_000;
const TOTAL_DISK_SPACE: u64 = 70_000_000;
const UPDATE_SIZE: u64 = 30_000_000;

fn main() {
    let input = aoc::read_input("2022/input_07.txt");
    let mut dir_sizes: HashMap<&str, u64> = HashMap::new();
    let mut path = vec!["/"];
    for line in input.lines() {
        if let Some(dir) = line.strip_prefix("$ cd ") {
            match dir {
                "/" => path.truncate(1),
                ".." => drop(path.pop()),
                _ => path.push(dir),
            }
        } else if line.starts_with(|c: char| c.is_ascii_digit()) {
            let (size_str, _) = line.split_once(' ').unwrap();
            let file_size: u64 = size_str.parse().unwrap();
            for dir in &path {
                *dir_sizes.entry(dir).or_default() += file_size;
            }
        }
    }
    let small_sum: u64 = dir_sizes.values().filter(|&&v| v <= DIR_MAX_SIZE).sum();
    println!("Part 1: {small_sum}");
    let free_space = TOTAL_DISK_SPACE - dir_sizes["/"];
    let min_to_free = UPDATE_SIZE - free_space;
    let dir_size = dir_sizes.values().filter(|&&v| v >= min_to_free).min().unwrap();
    println!("Part 2: {dir_size}");
}
